README
CHI-Driver-SharedMem Version 0.13

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc CHI::Driver::SharedMem

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=CHI-Driver-SharedMem

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/CHI-Driver-SharedMem

    CPAN Ratings
        http://cpanratings.perl.org/d/CHI-Driver-SharedMem

    Search CPAN
        http://search.cpan.org/dist/CHI-Driver-SharedMem/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Nigel Horne

This program is released under the following licence: GPL


Changes
Revision history for CHI::Driver::SharedMem

0.13	Mon Sep 17 09:20:58 BST 2012
	Handle race condition which could result in an error when two processes
		try to remove the same empty shared memory area when they
		terminate

0.12	Mon Sep  3 08:22:39 BST 2012
	t/01-new.t: Handle 'Bad System Call' within Cygwin
	Use read locks where possible

0.11	Fri Aug 31 08:31:20 BST 2012
	Use flock instead of IPC::Semaphore::Concurrency
	if($value) corrected to if(defined($value))

0.10	Wed Aug 22 13:53:49 BST 2012
	Use ftok to generate the shmkey in the test
	Don't try to build on GNU/Hurd
	Lock the shared memory area during accesses

0.09	Sat Aug 18 17:57:40 BST 2012
	Use $Config{intsize} do determine the size of an int
	Croak if the shared memory area can't be created

0.08	Wed Aug 15 08:31:38 BST 2012
	Added test that the shared memory area is deleted on cleanup
	Ensure using the latest version of CHI::Stats
	Don't set the execute bit in the shared memory area
	Store the size as an int rather than 32 bits

0.07	Mon Aug 13 09:04:17 BST 2012
	Attach to the shared memory area
	Only remove the shared memory area in the destructor when there are
		no attachments, as well as nothing in the cache
	Croak rather than carp when shmkey isn't given
	Tidied the tests

0.06	Sun Aug 12 20:17:56 BST 2012
	Don't try to remove the shared memory area if we haven't attached to it

0.05	Sun Aug 12 14:09:19 BST 2012
	Add consistency check which handles the failure at
		http://www.cpantesters.org/cpan/report/a7449b64-e424-11e1-9066-3d54fb7543f5
	Fixed the failing tests (test_mirror_cache and test_li_cache)

=head2 test_mirror_cache

0.04	Sat Aug 11 14:27:19 BST 2012
	Furged some tests to pass when they should fail, it's a temporary
		work around which will nee to be addressed

0.03	Fri Aug 10 18:07:19 BST 2012
	When the last object goes, if the cache is empty reclaim the shared
		memory
	More moosified, so it's much smaller and cleaner

0.02	Fri Aug 10 16:42:15 BST 2012
	Fixed some tests

0.01
        First draft


lib/CHI/Driver/SharedMem/t/CHIDriverTests.pm
package CHI::Driver::SharedMem::t::CHIDriverTests;

use strict;
use warnings;
use CHI::Test;
use base qw(CHI::t::Driver);
use Test::Warn;
use IPC::SysV;

use CHI::Test::Util
  qw(activate_test_logger cmp_bool is_between random_string skip_until);

=head1 NAME

CHI::Driver::SharedMem::t::CHIDriverTests

=head1 VERSION

Version 0.13

=cut

our $VERSION = '0.13';

=head1 SYNOPSIS

CHI::Driver::SharedMem::t::CHIDriverTests - test CHI::Driver::SharedMem

=cut

=head1 SUBROUTINES/METHODS

=head2 testing_driver_class

Declare the driver being tested

=cut

sub testing_driver_class {
	'CHI::Driver::SharedMem'
}

=head2 new_cache_options

=cut
sub new_cache_options {
	my $self = shift;

	return (
	    $self->SUPER::new_cache_options(),
	    driver => '+CHI::Driver::SharedMem',
	    size => 16 * 1024,
	    shmkey => IPC::SysV::ftok($0),	# hope it's unique :-(
	);
}

=head2 test_shmkey_required

Verify that the shmkey option is mandatory

=cut

sub test_shmkey_required : Tests {
	my $cache;

	eval {
		$cache = CHI->new(driver => 'SharedMem');
	};
	if($@) {
		ok($@ =~ /CHI::Driver::SharedMem - no key given/);
		ok(!defined($cache));
	} else {
		ok(0, 'Allowed shmkey to be undefined');
	}
}

1;


Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

if($^O eq 'MSWin32') {
	# This will set the CPAN-Testers status to Unknown
	# TODO: Work out how to set the status to NA
	die 'Windows is not supported';
} elsif($^O eq 'gnu') {
	# I don't believe that the hurd supports shared memory
	die 'GNU/Hurd is not supported';
}

WriteMakefile(
    NAME                => 'CHI::Driver::SharedMem',
    AUTHOR              => q{Nigel Horne <njh@bandsman.co.uk>},
    VERSION_FROM        => 'lib/CHI/Driver/SharedMem.pm',
    ABSTRACT_FROM       => 'lib/CHI/Driver/SharedMem.pm',
    ((defined($ExtUtils::MakeMaker::VERSION) &&
     ($ExtUtils::MakeMaker::VERSION >= 6.3002))
      ? ('LICENSE'=> 'GPL')
      : ()),
    PL_FILES            => {},
    PREREQ_PM => {
        'Test::More' => 0.82,	# Needs new_ok
	'Test::NoWarnings' => 0,
	'Test::Warn' => 0,
	'Test::Class' => 0,
	'Storable' => 0,
	'Moose' => 0,
	'IPC::SysV' => 0,
	'IPC::SharedMem' => 0,
	'CHI' => 0.54,	# Needs latest CHI::Stats
	'Config' => 0,
	'Fcntl' => 0,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'CHI-Driver-SharedMem-*' },
);


lib/CHI/Driver/SharedMem.pm
package CHI::Driver::SharedMem;

# There is an argument for mapping namespaces into keys and then putting
# different namespaces into different shared memory areas.  I will think about
# that.

use warnings;
use strict;
use Moose;
use IPC::SysV qw(S_IRUSR S_IWUSR IPC_CREAT);
use IPC::SharedMem;
use Storable qw(freeze thaw);
use Data::Dumper;
use Digest::MD5;
use Carp;
use Config;
use Fcntl;

extends 'CHI::Driver';

has 'shmkey' => (is => 'ro', isa => 'Int');
has 'shm' => (is => 'ro', builder => '_build_shm', lazy => 1);
has 'size' => (is => 'ro', isa => 'Int', default => 8 * 1024);
has 'lock' => (
	is => 'ro',
	builder => '_build_lock',
);
has '_data_size' => (
	is => 'rw',
	isa => 'Int',
	reader => '_get_data_size',
	writer => '_set_data_size'
);
has '_data' => (
	is => 'rw',
	isa => 'ArrayRef[ArrayRef]',
	reader => '_get_data',
	writer => '_set_data'
);

__PACKAGE__->meta->make_immutable();

=head1 NAME

CHI::Driver::SharedMem - Cache data in shared memory

=head1 VERSION

Version 0.13

=cut

our $VERSION = '0.13';

# FIXME - get the pod documentation right so that the layout of the memory
# area looks correct in the man page

=head1 SYNOPSIS

L<CHI> driver which stores data in shared memory objects for persistently over
processes.
Size is an optional parameter containing the size of the shared memory area, in
bytes.
Shmkey is a mandatory parameter containing the IPC key for the shared memory
area. See L<IPC::SharedMem> for more information.

    use CHI;
    my $cache = CHI->new(
	driver => 'SharedMem',
	size => 8 * 1024,
	shmkey => 12344321,	# Choose something unique
    );
    # ...

The shared memory area is stored thus:

	# Number of bytes in the cache [ int ]
	'cache' => {
		'namespace1' => {
			'key1' => 'value1',
			'key2' => 'value2',
			# ...
		},
		'namespace2' => {
			'key1' => 'value3',
			'key3' => 'value2',
			# ...
		}
		# ...
	}

=head1 SUBROUTINES/METHODS

=head2 store

Stores an object in the cache

=cut

sub store {
	my($self, $key, $value) = @_;

	$self->_lock(type => 'write');
	my $h = $self->_data();
	$h->{$self->namespace()}->{$key} = $value;
	$self->_data($h);
	$self->_unlock();
}

=head2 fetch

Retrieves an object from the cache

=cut

sub fetch {
	my($self, $key) = @_;

	$self->_lock(type => 'read');
	my $rc = $self->_data()->{$self->namespace()}->{$key};
	$self->_unlock();
	return $rc;
}

=head2 remove

Remove an object from the cache

=cut

sub remove {
	my($self, $key) = @_;

	$self->_lock(type => 'write');
	my $h = $self->_data();
	delete $h->{$self->namespace()}->{$key};
	$self->_data($h);
	$self->_unlock();
}

=head2 clear

Removes all data from the current namespace

=cut

sub clear {
	my $self = shift;

	$self->_lock(type => 'write');
	my $h = $self->_data();
	delete $h->{$self->namespace()};
	$self->_data($h);
	$self->_unlock();
}

=head2 get_keys

Gets a list of the keys in the current namespace

=cut

sub get_keys {
	my $self = shift;

	$self->_lock(type => 'read');
	my $h = $self->_data();
	$self->_unlock();
	return(keys(%{$h->{$self->namespace()}}));
}

=head2 get_namespaces

Gets a list of the namespaces in the cache

=cut

sub get_namespaces {
	my $self = shift;

	$self->_lock(type => 'read');
	my $rc = $self->_data();
	$self->_unlock();
	return keys(%{$rc});
}

# Internal routines

# The area must be locked by the caller
sub _build_shm {
	my $self = shift;

	my $shm = IPC::SharedMem->new($self->shmkey(), $self->size(), S_IRUSR|S_IWUSR);
	unless($shm) {
		$shm = IPC::SharedMem->new($self->shmkey(), $self->size(), S_IRUSR|S_IWUSR|IPC_CREAT);
		unless($shm) {
			croak 'Couldn\'t create a shared memory area with key ' .
				$self->shmkey();
			return;
		}
		$shm->write(pack('I', 0), 0, $Config{intsize});
	}
	$shm->attach();
	return $shm;
}

sub _build_lock {
	open(my $fd, '<', $0);
	return $fd;
}

sub _lock {
	my ($self, %params) = @_;

	flock($self->lock(), ($params{type} eq 'read') ? Fcntl::LOCK_SH : Fcntl::LOCK_EX);
}

sub _unlock {
	my $self = shift;

	flock($self->lock(), Fcntl::LOCK_UN);
}

# The area must be locked by the caller
sub _data_size {
	my($self, $value) = @_;

	if(defined($value)) {
		$self->shm()->write(pack('I', $value), 0, $Config{intsize});
		return $value;
	}
	unless($self->shm()) {
		return 0;
	}
	return unpack('I', $self->shm()->read(0, $Config{intsize}));
}

# The area must be locked by the caller
sub _data {
	my($self, $h) = @_;

	if(defined($h)) {
		my $f = freeze($h);
		my $cur_size = length($f);
		$self->shm()->write($f, $Config{intsize}, $cur_size);
		$self->_data_size($cur_size);
		return $h;
	}
	my $cur_size = $self->_data_size();
	unless($cur_size) {
		return {};
	}
	return thaw($self->shm()->read($Config{intsize}, $cur_size));
}

=head2 BUILD

Constructor - validate arguments

=cut

sub BUILD {
	my $self = shift;

	unless($self->shmkey()) {
		croak 'CHI::Driver::SharedMem - no key given';
	}
}

=head2 DEMOLISH

If there is no data in the shared memory area, and no-one else is using it,
it's safe to remove it and reclaim the memory.

=cut

sub DEMOLISH {
	my $self = shift;

	if($self->shmkey()) {
		my $cur_size;
		if(scalar($self->get_namespaces())) {
			$self->_lock(type => 'read');
			$cur_size = $self->_data_size();
			$self->_unlock();
		} else {
			$self->_lock(type => 'write');
			$self->_data_size(0);
			$self->_unlock();
			$cur_size = 0;
		}
		$self->shm()->detach();
		# We could scan the cache and see if all has expired.
		# If it has, then the cache could be removed if nattch = 0.
		unless($cur_size) {
			my $stat = $self->shm()->stat();
			if(defined($stat) && ($stat->nattch() == 0)) {
				$self->shm()->remove();
			}
		}
	}
}

=head1 AUTHOR

Nigel Horne, C<< <njh at bandsman.co.uk> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-chi-driver-sharedmem at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=CHI-Driver-SharedMem>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.


=head1 SEE ALSO

CHI, IPC::SharedMem


=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc CHI::Driver::SharedMem


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=CHI-Driver-SharedMem>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/CHI-Driver-SharedMem-Info>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/CHI-Driver-SharedMem>

=item * Search CPAN

L<http://search.cpan.org/dist/CHI-Driver-SharedMem>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 Nigel Horne.

This program is released under the following licence: GPL


=cut

1; # End of CHI::Driver::SharedMem
